tests/unit/APNSResponseTest.php

```diff
@@ -19,4 +19,31 @@ public function testThatResponseParserReadsErrorResponse() {
 		$this->assertEquals( 400, $response->getStatusCode() );
 		$this->assertEquals( 'BadDeviceToken', $response->getErrorMessage() );
 	}
+
+	public function testUnrecoverableErrorEvaluation() {
+		$this->assertTrue( $this->makeAPNSResponseFor( 400 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 403 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 404 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 405 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 410 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 413 )->isUnrecoverableError() );
+		$this->assertFalse( $this->makeAPNSResponseFor( 429 )->isUnrecoverableError() );
```

Comment:
This test doesn't subscribe to the "one (logical) assertion per test" rule, as it tests different input-output combos.

I think it's acceptable to have it like this, though, because the behavior it tests is the overall logic around which errors are unrecoverable or not. Having one test function per status code would result in more noise, imho.

If this is against the coding conventions for the repo / PHP testing world, I'm happy to change it, though.

Reply:
I don't mind doing multiple assertions at all – in this case it definitely makes sense. 

One thing I would suggest – the test names are typically meant to be an explanation of the behaviour the test is validating (ie – `testThat...`). Ideally we'd keep that up, though in some cases it _can_ be a little awkward – the main benefit here is that it hopefully helps explain the _why_ rather than the _what_, especially in a test like this that sort of tests that the code does what the code does.

One suggestion: `testThatUnrecoverableErrorsAreCorrectlyRecognized`, but I'll leave it to you to decide what's best :)

Reply:
Oh, right, the name 🤦 I did notice the "test that ..." in my previous reviews, but somehow I missed it in my own writing 😳 .

Updated.